File: source/javascript/thread_safe_queue.js
// Thread-Safe Queue (async simulation) - Multiple producers, single consumer
async function producer(id, items, count) {
    const result = [];
    for (let i = 0; i < count; i++) result.push(`p${id}-${i}`);
    return result;
}

async function main() {
    const numProducers = 3;
    const itemsPerProducer = 5;
    const totalItems = numProducers * itemsPerProducer;

    const producerResults = await Promise.all(
        Array.from({ length: numProducers }, (_, i) => producer(i, [], itemsPerProducer))
    );

    const results = producerResults.flat();
    console.log(`consumed ${results.length} items`);
    results.sort();
    console.log(`values: ${results.join(" ")}`);
}

main();

File: source/go/thread_safe_queue.go
package main

import (
	"fmt"
	"sort"
	"strings"
	"sync"
)

func producer(id int, queue chan<- string, count int, wg *sync.WaitGroup) {
	defer wg.Done()
	for i := 0; i < count; i++ {
		queue <- fmt.Sprintf("p%d-%d", id, i)
	}
}

func consumer(queue <-chan string, expectedCount int) []string {
	var received []string
	for i := 0; i < expectedCount; i++ {
		val := <-queue
		received = append(received, val)
	}
	return received
}

func main() {
	queue := make(chan string, 100)
	numProducers := 3
	itemsPerProducer := 5
	totalItems := numProducers * itemsPerProducer

	var wg sync.WaitGroup
	resultCh := make(chan []string, 1)

	go func() {
		resultCh <- consumer(queue, totalItems)
	}()

	for i := 0; i < numProducers; i++ {
		wg.Add(1)
		go producer(i, queue, itemsPerProducer, &wg)
	}

	wg.Wait()

	results := <-resultCh
	close(queue)

	fmt.Printf("consumed %d items\n", len(results))
	sort.Strings(results)
	fmt.Printf("values: %s\n", strings.Join(results, " "))
}
